### PR TITLE
Handle compiler diagnostics reported in .cshtml files during build

### PR DESCRIPTION
**Pull request by mavasani**
Fixes [AB#1511940](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1511940)

Compiler errors/warnings reported in .cshtml files during build were not being reported in the error list, but reported just in the output window if the .cshtml files were not open in VS. If the user opens these files in VS after the build, intellisense kicks in and reports these diagnostics on the generated .ide.g.cs files with a mapped location in the .cshtml files.

In order to ensure that these build diagnostics in .cshtml are reported in error list for closed .cshtml files, we now classify these as build-only additional file diagnostics. I also fixed couple of places in the code path where we were not handling diagnostics from build being reported on non-source text documents. I verified that after these fixes, the repro in the above feedback ticket is fixed.

One other alternative approach to fix this issue would be to not classify these .cshtml file diagnostics from build as build-only diagnostics, but instead perform the reverse secondary buffer mapping as done by `VisualStudioVenusSpanMappingService` here: https://github.com/dotnet/roslyn/blob/4298934df345f501fa0a74b675fab8efd8355709/src/VisualStudio/Core/Def/Diagnostics/VisualStudioVenusSpanMappingService.cs#L29-L31 and convert it to live diagnostic with primary location in the backing .ide.g.cs file and secondary mapped location in the .cshtml file, and this should ensure correct build-live diagnostic de-deduping. However, this seems to be non-trivial amount of work, and certainly not a bug fix level change. I'll let the razor team decide if they want to explore this route, and if so extend the `VisualStudioVenusSpanMappingService` to support this reverse mapping.

### Experience before fix
CshtmlErrors_BeforeFix

### Experience after fix
CshtmlErrors_AfterFix

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @arkalyanms - do we want to consider this fix for 17.5? If so, I can retarget the fix to 17.5 release branch.

File: src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 1]
```diff
@@ -797,6 +797,15 @@ private bool IsLive(Project project, DiagnosticData diagnosticData)
                     return false;
                 }
 
+                // Compiler diagnostics reported on additional documents indicate mapped diagnostics, such as compiler diagnostics
+                // in razor files which are actually reported on generated source files but mapped to razor files during build.
+                // These are not reported on additional files during live analysis, and can be considered to be build-only diagnostics.
+                if (IsAdditionalDocumentDiagnostic(project, diagnosticData) &&
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs:803]: This is the core fix to classify these compiler build diagnostics in .cshtml files as build-only diagnostics.

**Review by davidwengier — APPROVED** [review]
I didn't know this was broken, have no idea how this whole system works, and therefore it is my expert opinion that this is fine.

**Comment by mavasani** [comment]
> no idea how this whole system works

Shhh. I don't know anyone who does! One more reason to love LSP, it will no longer be our problem to hide :-)

**Review by genlu — APPROVED** [review]
:shipit:

**Comment by arkalyanms** [comment]
> Tagging @arkalyanms - do we want to consider this fix for 17.5? If so, I can retarget the fix to 17.5 release branch.

does this impact all razor projects? If so, yes.

**Comment by mavasani** [comment]
> oes this impact all razor projects? If so, yes.

Yes, it does. I'll retarget

**Comment by mavasani** [comment]
@arkalyanms This now targets release/dev17.5 branch. Please feel free to merge whenever required approvals have been taken.

**Comment by arkalyanms** [comment]
@davidwengier would this be a test we can include in Bala's manual suite?

**Comment by davidwengier** [comment]
It wouldn't hurt, but its also something that it should be possible to create an integration test for it in this repo too, to know ASAP if it breaks.

**Comment by arkalyanms** [comment]
> It wouldn't hurt, but its also something that it should be possible to create an integration test for it in this repo too, to know ASAP if it breaks.

@mavasani can we add an integration test for this? Not a part of this PR.

**Comment by mavasani** [comment]
> > It wouldn't hurt, but its also something that it should be possible to create an integration test for it in this repo too, to know ASAP if it breaks.
> 
> @mavasani can we add an integration test for this? Not a part of this PR.

Do we have Razor integration tests in Roslyn? If so, can someone point me to them>

**Comment by davidwengier** [comment]
Probably not specifically, but should be able to create one in the same way as existing tests. ie, [Create a project](https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpNewDocumentFormatting.cs#L51), open a file, cause a compiler error, close the file, verify the contents of the error list.

**Comment by mavasani** [comment]
> Probably not specifically, but should be able to create one in the same way as existing tests. ie, [Create a project](https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpNewDocumentFormatting.cs#L51), open a file, cause a compiler error, close the file, verify the contents of the error list.

Thanks, let me give it a shot today.

**Comment by mavasani** [comment]
> > Probably not specifically, but should be able to create one in the same way as existing tests. ie, [Create a project](https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpNewDocumentFormatting.cs#L51), open a file, cause a compiler error, close the file, verify the contents of the error list.
> 
> Thanks, let me give it a shot today.

FYI: I did give this a try, but was unable to cross the bridge here. There are a bunch of web templates that ship, and I was able to add them to our integration test supported templates, but all the templates bring up a wizard to select project template parameters to configure the kind of project to generate. I tried bunch of approaches and APIs and also passing in custom template parameters, but none of them seem to work. I have pushed a branch/commit here in case someone else wants to give it a try: https://github.com/dotnet/roslyn/commit/fe9f967929979d15bc6c75612e9673f2221960bd

**Comment by davidwengier** [comment]
The Blazor Server project template doesn't have any additional popups, so should work better. Will see if I can get something going in https://github.com/dotnet/roslyn/pull/66459

Thanks for the head start!

**Pull request merged**